### PR TITLE
Add AST module caching support for LSP

### LINKED ISSUE #4994: Implement new allocator and support garbage collection
For supporting long-term compilation sessions (for LSP use) we want to support garbage collecting the engines.

Right now this is not possible due to our current slab allocator not supporting removing previously allocated ids.

We should re-implement the slab allocator using either an optimized slot map, or the buddy allocator algorithm.

(Tests using a slotmap crate based implementation proved to be 15-20% slower than the current allocator.)

Additionally we can move to a multiple allocator model, where we setup a slab per module, and allow cross-linking of type/decl ids.

This will also allow for easier incremental compilation as well as less transient caching due to a more straight forward cache serialization.

### PR DESCRIPTION
**Pull request by JoshuaBatty**
## Description
This PR supersedes #4988 and builds on backend work done by @tritao in #4733.

A new `ModuleId` is created for each compiled module. The `SourceId` type now also stores the `ModuleId` that was used to create it. This allows us to perform garbage collection on types in the Declaration and Type engines that refer the provided module.

```rust
if let Some(module_id) = engines.se().get_module_id(&path) {
    engines.clear_module(&module_id);
}
```

This method isn't super cheap to call. My tests had this averaging at 54ms. We are currently performing garbage collection on every key stroke in LSP. We should look to have another metric such as only clearing once every 20 keystrokes or so in the future.

I ran benchmarking on recompiling the `doc_comments` example 20 times to simulate rapid keystrokes during coding. Here are the results.

| Without Caching | With Caching |
|---------------------|------------------|
| 89.489 s             | 4.6236  s         |

A speed up of 1835.49% is certainly great, yet 4.6 seconds still sounds like an eternity. Still have lots of room to improve but this should improve our current system dramatically. 

closes #4994

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [ ] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by IGI-111** [comment]
Very nice. It probably needs to be tuned further, but it's a great start and much cleaner than I would have expected for something like this.

**Comment by JoshuaBatty** [comment]
Great, I'll start think about a nice heuristic to call gc from.

File: sway-lsp/tests/lib.rs [hunk 1]
```diff
@@ -95,6 +95,43 @@ async fn did_change() {
...
+#[tokio::test]
+async fn did_cache_test() {
+    let (mut service, _) = LspService::build(ServerState::new)
+        .custom_method("sway/metrics", ServerState::metrics)
+        .finish();
+    let uri = init_and_open(&mut service, doc_comments_dir().join("src/main.sw")).await;
+    let _ = lsp::did_change_request(&mut service, &uri).await;
+    let metrics = lsp::metrics_request(&mut service, &uri).await;
+    assert!(metrics.len() >= 2);
+    println!("metrics req: {:?}", metrics);
```
Comment (sdankel) [inline sway-lsp/tests/lib.rs:107]: nit

```suggestion
```

File: sway-lsp/src/handlers/request.rs [hunk 2]
```diff
@@ -447,3 +448,33 @@ pub fn handle_visualize(
...
+        Ok((_, session)) => {
+            let engines = session.engines.read();
+            let mut metrics = vec![];
+            for kv in session.metrics.iter() {
+                let path = engines
+                    .se()
+                    .get_path(kv.key())
+                    .to_str()
+                    .unwrap_or("")
+                    .to_string();
```
Comment (sdankel) [inline sway-lsp/src/handlers/request.rs:470]: nit:
```suggestion
                    .to_string_lossy();
```

**Review by sdankel — DISMISSED** [review]
Amazing performance improvement!

File: sway-types/Cargo.toml [hunk 3]
```diff
@@ -16,6 +16,7 @@ lazy_static = "1.4"
 num-bigint = "0.4.3"
 num-traits = "0.2.16"
 serde = { version = "1.0", features = ["derive"] }
+sway-utils = { version = "0.46.1", path = "../sway-utils" }
```
[the remark this answers is not in the document]
Comment (JoshuaBatty) [reply sway-types/Cargo.toml:19]: We are using it in the source engine.

```rust
let manifest_path = sway_utils::find_parent_manifest_dir(path).unwrap_or(path.clone());
```

File: sway-core/src/type_system/info.rs [hunk 4]
```diff
@@ -67,6 +67,28 @@ impl<T: PartialEqWithEngines> PartialEqWithEngines for VecSet<T> {
     }
 }
 
+/// Encapsulates type information and its optional source identifier.
+#[derive(Debug, Default, Clone)]
+pub struct TypeData {
```
[the remark this answers is not in the document]
Comment (JoshuaBatty) [reply sway-core/src/type_system/info.rs:72]: Nice, i'll change it to this. Was thinking this could have a better name. Thanks.

**Comment by JoshuaBatty** [comment]
> Overall looks great, though it seems some of the type engine insertions have a `None` source id. As far as I can tell, that means we won't track those, so we won't be able to GC them, right? Assuming that's the case, something we should look into improving in the future.

Good question, at the top of the type engines `insert` function we check if `None` was provided

```rust
let source_id = source_id.map(Clone::clone).or_else(|| info_to_source_id(&ty));
``` 

If it's None then we try to map the type_info to a reserved source id. 

```rust
/// Maps specific `TypeInfo` variants to a reserved `SourceId`, returning `None` for non-mapped types.
fn info_to_source_id(ty: &TypeInfo) -> Option<SourceId> {
    match ty {
        TypeInfo::Unknown
        | TypeInfo::UnsignedInteger(_)
        | TypeInfo::Numeric
        | TypeInfo::Boolean
        | TypeInfo::B256
        | TypeInfo::RawUntypedPtr
        | TypeInfo::RawUntypedSlice
        | TypeInfo::StringSlice
        | TypeInfo::Contract
        | TypeInfo::StringArray(_)
        | TypeInfo::Array(_, _) => Some(SourceId::reserved()),
        TypeInfo::Tuple(v) if v.is_empty() => Some(SourceId::reserved()),
        _ => None,
    }
}
```

This way we don't clear types that are shared between modules, and type_info variants that can hold spans are allowed to be None, thus can be cleared safely.

**Review by sdankel — DISMISSED** [review]

**Comment by JoshuaBatty** [comment]
> Calling it on every keystroke certainly seems excessive, even on every couple keystrokes. I think it could be worth thinking about having a memory size heuristics-based system instead, and calling it on idle time so it's less perceptible to the user.

I'm now only calling it on every 10th keystroke. Unfortunately we can't call it during idle and this would remove internal types needed for doing goto def, hover etc in the users workspace.

**Review by sdankel — APPROVED** [review]

**Comment by JoshuaBatty** [comment]
hmm calling it on every 10th keystroke is causing CI to run out of memory. I'll open an issue and we can work on this optimisation in a follow up PR.

**Review by sdankel — APPROVED** [review]

File: sway-lsp/tests/lib.rs [hunk 5]
```diff
@@ -112,25 +112,6 @@ async fn did_cache_test() {
     shutdown_and_exit(&mut service).await;
 }
 
-#[tokio::test]
```
[the remark this answers is not in the document]
Comment (JoshuaBatty) [reply sway-lsp/tests/lib.rs]: Sure that's a good idea i'll update it, thanks.

**Review by sdankel — APPROVED** [review]

**Pull request merged**